File: Assets/Game/Scripts/Moduls/P_Modul_26.lua
local ObjectStateBackupUtils = Ballance2.Utils.ObjectStateBackupUtils

---@class P_Modul_26 : ModulBase 
---@field P_Modul_26_Rope PhysicsObject
---@field P_Modul_26_Sack PhysicsObject
---@field P_Modul_26_Sack_Force PhysicsConstraintForce
P_Modul_26 = ModulBase:extend()

function P_Modul_26:new()
  P_Modul_26.super.new(self)
  self._ForceTimer = nil
  self._ForceState = false
end
function P_Modul_26:Active()
  ModulBase.Active(self)
  self.P_Modul_26_Rope:Physicalize()
  self.P_Modul_26_Sack:Physicalize()
  self.P_Modul_26_Sack.EnableConstantForce = true
  self._ForceTimer = LuaTimer.Add(500, 1500, function ()
    self._ForceState = not self._ForceState

    --切换方向
    if self._ForceState then
      self.P_Modul_26_Sack_Force.Force = 0.5
    else
      self.P_Modul_26_Sack_Force.Force = -0.5
    end
  end)
end
function P_Modul_26:Deactive()
  self.P_Modul_26_Sack.EnableConstantForce = false
  self.P_Modul_26_Rope:UnPhysicalize(true)
  self.P_Modul_26_Sack:UnPhysicalize(true)
  if self._ForceTimer then
    LuaTimer.Delete(self._ForceTimer)
    self._ForceTimer = nil
  end
  ModulBase.Deactive(self)
end

function P_Modul_26:ActiveForPreview()
  self.gameObject:SetActive(true)
end
function P_Modul_26:DeactiveForPreview()
  self.gameObject:SetActive(false)
end

function P_Modul_26:Reset()
  ObjectStateBackupUtils.RestoreObject(self.P_Modul_26_Rope.gameObject)
  ObjectStateBackupUtils.RestoreObject(self.P_Modul_26_Sack.gameObject)
  self._ForceIsLeft = false
  self._ForceTick = 0
end
function P_Modul_26:Backup()
  ObjectStateBackupUtils.BackUpObject(self.P_Modul_26_Rope.gameObject)
  ObjectStateBackupUtils.BackUpObject(self.P_Modul_26_Sack.gameObject)
end

function CreateClass:P_Modul_26()
  return P_Modul_26()
end

File: Assets/Game/GamePlay/Scripts/Moduls/P_Modul_26.cs
using Ballance2.Services;
using Ballance2.Utils;
using BallancePhysics.Wapper;

namespace Ballance2.Game.GamePlay.Moduls
{
  public class P_Modul_26 : ModulBase 
  {
    public PhysicsObject P_Modul_26_Sack;
    public PhysicsObject P_Modul_26_Rope;
    public PhysicsConstraintForce P_Modul_26_Sack_Force;
    public float Force = 0.25f;
    public float SwitchTime = 1.400f;

    private int  _ForceTimer = 0;
    private bool _ForceState = true;

    public P_Modul_26() {
      EnableBallRangeChecker = true;
      BallCheckeRange = 30;
    }

    public override void Active()
    {
      base.Active();
      P_Modul_26_Rope.gameObject.SetActive(true);
      P_Modul_26_Sack.gameObject.SetActive(true);
      P_Modul_26_Rope.Physicalize();
      P_Modul_26_Sack.Physicalize();
      P_Modul_26_Sack.EnableConstantForce = true;
      _ForceTimer = GameTimer.Timer(SwitchTime, () => {
        _ForceState = !_ForceState;//切换方向
        if (_ForceState)
          P_Modul_26_Sack_Force.Force = Force;
        else
          P_Modul_26_Sack_Force.Force = -Force;
      });
    }

    public override void Deactive()
    {
      P_Modul_26_Sack.EnableConstantForce = false;
      P_Modul_26_Rope.UnPhysicalize(true);
      P_Modul_26_Sack.UnPhysicalize(true);
      if (_ForceTimer > 0) {
        GameTimer.DeleteTimer(_ForceTimer);
        _ForceTimer = 0;
      }
      base.Deactive();
    }

    public override void ActiveForPreview()
    {
      gameObject.SetActive(true);
    }

    public override void DeactiveForPreview()
    {
      gameObject.SetActive(false);
    }

    public override void Reset(ModulBaseResetType type)
    {
      ObjectStateBackupUtils.RestoreObject(P_Modul_26_Rope.gameObject);
      ObjectStateBackupUtils.RestoreObject(P_Modul_26_Sack.gameObject);
    }

    public override void Backup()
    {
      ObjectStateBackupUtils.BackUpObject(P_Modul_26_Rope.gameObject);
      ObjectStateBackupUtils.BackUpObject(P_Modul_26_Sack.gameObject);
    }
  }
}